### PR TITLE
Implement Workitem support in TRX logger

### PR DESCRIPTION
**Pull request by fl3pp**
## Description

The former `MSTest.exe` created TRX-files which may contain `Workitem` attributes:

``` XML
...
<UnitTest name="%test_name%" storage="%path_to_dll%" id="7212e5d1-c0ef-d4a3-55b9-4dbcf7611dd6">
  <TestCategory>
	<TestCategoryItem TestCategory="%category%" />
  </TestCategory>
  <Execution id="23e0a237-290e-491d-a858-d8aae2e7ee04" />
  <Workitems>
	<Workitem>%work_item_id%</Workitem>
  </Workitems>
  <TestMethod codeBase="%path_to_dll%" adapterTypeName="%adapter_dll%" className="%test_class_name%" name="%test_method_name%" />
</UnitTest>
...
```

Visual Studio is still able to read this information in its TRX-viewer:

vs-trx-viewer

This PR re-adds support for Workitems in TRX-files.

## Related issue
- While MSTestV2 has an implementation for Workitems, there is a bug so that Workitems are not passed on to the test platform. This [PR](https://github.com/microsoft/testfx/pull/749) fixes this bug

### CODE REVIEW

**Comment by jflepp** [comment]
It seems like there's an issue with the payment methods of the artifact storage. The build and the tests passed: [build](https://dev.azure.com/vstestplatform/TestPlatform/_build/results?buildId=2012&view=logs&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=1671bef0-58cd-5026-1035-b95b47d397bd)

**Comment by Haplois** [comment]
We're addressing the build pipeline issue thanks for the heads up.

This PR contains the same functionality in https://github.com/microsoft/vstest/pull/1801; we need to review both and be careful about which one to merge.

**Comment by jflepp** [comment]
The PR #1801 does make sense, but the Workitem implementation is not compatible with Visual Studio and there hasn't been any activity in the PR since almost _two years_.

We've seen major workarounds around this issue (like manually re-scanning all test assemblies) and think it makes sense to solve this issue in the right place, as we try to go forward with a new project dependent on this information.

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/Interfaces/ITestElement.cs [hunk 1]
```diff
@@ -16,6 +16,7 @@ internal interface ITestElement
         TestExecId ParentExecutionId { get; set; }
         TestListCategoryId CategoryId { get; set; }
         TestCategoryItemCollection TestCategories { get; }
+        WorkitemCollection Workitems { get; set; }
```
Comment (Haplois) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/Interfaces/ITestElement.cs:19]: Please match the case with the attribute name. 

```suggestion
        WorkItemCollection WorkItems { get; set; }
```

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs [hunk 2]
```diff
@@ -30,6 +30,7 @@ internal abstract class TestElement : ITestElement, IXmlTestStore
         protected TestExecId executionId;
         protected TestExecId parentExecutionId;
         protected TestCategoryItemCollection testCategories;
+        protected WorkitemCollection workitems;
```
Comment (Haplois) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs:33]:
```suggestion
        protected WorkItemCollection workItems;
```

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs [hunk 3]
```diff
@@ -161,6 +162,20 @@ public TestCategoryItemCollection TestCategories
             }
         }
 
+        /// <summary>
+        /// Gets or sets the work items.
+        /// </summary>
+        public WorkitemCollection Workitems
```
Comment (Haplois) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs:168]: Same here, please make sure you match the case.

**Review by Haplois — CHANGES_REQUESTED** [review]
I tested the changes; it looks good. I know our attribute is named `WorkItemAttribute`, and in XML, we're using the form `Workitem`. But let's keep at least the code part is consistent. 

Can you update the casings? After that, it's good to merge.

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/Interfaces/ITestElement.cs (hunk 1 above)
Comment (jflepp) [reply src/Microsoft.TestPlatform.Extensions.TrxLogger/Interfaces/ITestElement.cs]: Updated naming

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs (hunk 2 above)
Comment (jflepp) [reply src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs]: Updated naming

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs (hunk 3 above)
Comment (jflepp) [reply src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs]: I updated all casings.

I had to make some more changes in order to keep the custom XML-serialization implementation happy. Unfortunately,  I will not be able to test these changes myself until _Thursday_, so I cannot promise it actually works.

I will keep you updated about the tests, thank you for the review!

Comment (Haplois) [reply src/Microsoft.TestPlatform.Extensions.TrxLogger/ObjectModel/TestElement.cs]: Yep. I'll verify the changes, and fix it if needed.

**Review by Haplois — APPROVED** [review]
Thank you for your contribution, I validated the changes, LTGM.

**Pull request merged**